### PR TITLE
Add cluster id check for hash kv handler

### PR DESCRIPTION
**Pull request by CaojiamingAlan**
See https://github.com/etcd-io/etcd/issues/15548
This PR only adds cluster id check for hashKVHandler. Cluster id checks for other handlers(member, promote, version, etc.) will be in separate PRs.

### CODE REVIEW

File: server/etcdserver/corrupt_test.go
Comment (serathius) [inline server/etcdserver/corrupt_test.go:482]: This is a first test for hashKVHandler on http layer, can you add more test cases? for example a case that the request succeeds so we not only test a failure.

File: server/etcdserver/corrupt_test.go
Comment (serathius) [inline server/etcdserver/corrupt_test.go:24]: Please add a test case with invalid cluster id to `TestInitialCheck`, `TestPeriodicCheck` and `TestCompactHashCheck`.

**Comment by serathius** [comment]
Change overall looks good, but would like to have more tests as this is pretty critical functionality. If possible I would also like to have an e2e tests for inplace cluster restore. Do you think you can add an e2e tests?

**Comment by CaojiamingAlan** [comment]
> Change overall looks good, but would like to have more tests as this is pretty critical functionality. If possible I would also like to have an e2e tests for inplace cluster restore. Do you think you can add an e2e tests?

Sure. Working on it.

File: tests/e2e/corrupt_test.go
Comment (jmhbnz) [inline tests/e2e/corrupt_test.go:160]: nit: typos
```suggestion
			t.Fatalf("There is no corruption after in-place recovery, but corruption reported.")
```

**Review by jmhbnz — COMMENTED** [review]
Thanks for your work on this @CaojiamingAlan.

Can you please squash the two identically named commits for `etcdserver: add e2e test for periodic check after in-place recovery`.

**Comment by CaojiamingAlan** [comment]
@serathius I've added the tests. Could you please review them? thx.

**Comment by chaochn47** [comment]
Hi @CaojiamingAlan I can help review the PR in case Marrek bandwidth is limited, give me a hour to refresh my mind on this area of code.

**Review by chaochn47 — APPROVED** [review]
LGTM with some nit-picks.

File: tests/e2e/corrupt_test.go
Comment (chaochn47) [inline tests/e2e/corrupt_test.go:101]: Nice test ~

**Comment by CaojiamingAlan** [comment]
@chaochn47 Comments fixed. Thank you for your review.

**Comment by chaochn47** [comment]
Thanks @CaojiamingAlan, please consider squash all the commits before asking for merging the PR~

**Comment by CaojiamingAlan** [comment]
@chaochn47 Squashed.

File: tests/e2e/corrupt_test.go
Comment (serathius) [reply tests/e2e/corrupt_test.go:101]: Can't agree more. Very nice work.

File: tests/e2e/corrupt_test.go
Comment (serathius) [inline tests/e2e/corrupt_test.go:118]: Can you change the comments into logs. When running this test I would like to see the current operation being executed. For example
```suggestion
	t.Log("Put some data into the old cluster, so that after recovering from a blank db, the hash diverges")
```

File: tests/e2e/corrupt_test.go
Comment (serathius) [inline tests/e2e/corrupt_test.go:149]: Part of the start time is waiting for the member to be ready. This causes the test to wait indefinitely and not really test the inplace restore. Please make sure that the test is able to do an inplace restore with corruption checking disabled and leads to corrupted state. And only then create a separate test that shows that you can prevent corruption by  enabling the checks.

Comment (CaojiamingAlan) [reply tests/e2e/corrupt_test.go:149]: I am a little bit confused.

>Part of the start time is waiting for the member to be ready. If I run the test without your fixes do corruption detection, will wait indefinitely for first member to join the cluster. 

Could you please double check or provide some log? I tried to run the test without my fix, and there was no indefinite wait. It reported corruptions as expected, and seemed the new cluster was on by log.

The new cluster (pid43115-43118)
```
{"Time":"2023-05-25T18:48:30.482422-05:00","Action":"output","Test":"TestInPlaceRecovery","Output":"/Users/caojiaming/go/src/etcd/bin/etcd (TestInPlaceRecovery-test-0) (43115): {\"level\":\"info\",\"ts\":\"2023-05-25T18:48:30.482351-0500\",\"logger\":\"raft\",\"caller\":\"v3@v3.0.0-20221201111702-eaa6808e1f7a/raft.go:797\",\"msg\":\"ca50e9357181d758 became leader at term 2\"}\r\n"}
...
{"Time":"2023-05-25T18:48:30.478547-05:00","Action":"output","Test":"TestInPlaceRecovery","Output":"/Users/caojiaming/go/src/etcd/bin/etcd (TestInPlaceRecovery-test-1) (43117): {\"level\":\"info\",\"ts\":\"2023-05-25T18:48:30.478433-0500\",\"logger\":\"raft\",\"caller\":\"v3@v3.0.0-20221201111702-eaa6808e1f7a/raft.go:728\",\"msg\":\"5ac8aa22f1eb4c8f became follower at term 2\"}\r\n"}
...
{"Time":"2023-05-25T18:48:37.571722-05:00","Action":"output","Test":"TestInPlaceRecovery","Output":"/Users/caojiaming/go/src/etcd/bin/etcd (TestInPlaceRecovery-test-2) (43118): {\"level\":\"info\",\"ts\":\"2023-05-25T18:48:37.571695-0500\",\"logger\":\"raft\",\"caller\":\"v3@v3.0.0-20221201111702-eaa6808e1f7a/raft.go:728\",\"msg\":\"72e86eb5e356a737 became follower at term 2\"}\r\n"}
```

and the old cluster is down (pid 43102-43104):
```
{"Time":"2023-05-25T18:48:30.420149-05:00","Action":"output","Test":"TestInPlaceRecovery","Output":"/Users/caojiaming/go/src/etcd/bin/etcd (TestInPlaceRecovery-test-2) (43104): {\"level\":\"warn\",\"ts\":\"2023-05-25T18:48:30.420045-0500\",\"logger\":\"raft\",\"caller\":\"v3@v3.0.0-20221201111702-eaa6808e1f7a/raft.go:1039\",\"msg\":\"665e70146a5ab3b6 stepped down to follower since quorum is not active\"}\r\n"}
```

> Please make sure that the test is able to do an inplace restore with corruption checking disabled and leads to corrupted state. 

Do you mean with the cluster-id check disabled? With corruption checking disabled we cannot even know there's a corruption. If so, should I introduce a command-line arg or environment variable or something that turns on/off the check?

Thank you for you patience in advance.

Comment (serathius) [reply tests/e2e/corrupt_test.go:149]: I think I cheated a little bit and used some older etcd binary. Possibly this works differently on current main branch. Give me some time to confirm.

Comment (CaojiamingAlan) [reply tests/e2e/corrupt_test.go:153]: @serathius Any updates? Can this be merged? Thanks！

Comment (serathius) [reply tests/e2e/corrupt_test.go:153]: LGTM

File: server/etcdserver/corrupt_test.go
Comment (serathius) [inline server/etcdserver/corrupt_test.go:535]: Please have explicit field names.

File: server/etcdserver/corrupt_test.go
Comment (serathius) [inline server/etcdserver/corrupt_test.go:536]: Instead of naming the test just "success", please describe how the success looks like as it might not be clear and change with time. For example "HashKV returns 200 if cluster hash matches" for success and ""HashKV returns 400 if cluster hash doesn't matche""

File: tests/e2e/corrupt_test.go
Comment (serathius) [inline tests/e2e/corrupt_test.go:172]: we should run `t.Cleanup` when creating a resource that needs cleanup. Not at the end. When tests fails it will exit early and this cleanup will not be run.

**Review by serathius — APPROVED** [review]
LGTM with some comments.

File: api/v3rpc/rpctypes/error.go
Comment (ahrtr) [inline api/v3rpc/rpctypes/error.go:50]: duplicated definition https://github.com/etcd-io/etcd/blob/693f25d6b98f5070527bc2b461f0992227c01c14/server/etcdserver/api/rafthttp/http.go#L57

Please raise a followup issue to reuse the same error. Also please also search "cluster ID mismatch" in the code base, there are also some hard code error messages. Please also consider to use the error.

File: server/etcdserver/corrupt.go
Comment (ahrtr) [inline server/etcdserver/corrupt.go:605]: It isn't good to compare a **hardcode** string. Please follow the same pattern as `mvcc.ErrCompacted`

File: api/v3rpc/rpctypes/error.go
Comment (jmhbnz) [reply api/v3rpc/rpctypes/error.go:50]: Note there is also a typo on `ErrGPRCIdMismatch`. 

```suggestion
	ErrGRPCIdMismatch             = status.Error(codes.FailedPrecondition, "etcdserver: cluster ID mismatch")
```

**Comment by ahrtr** [comment]
@CaojiamingAlan Have you raised a followup ticket for https://github.com/etcd-io/etcd/pull/15924#discussion_r1224894632? 

Please also rebase this PR although github doesn't show any conflict.

**Comment by CaojiamingAlan** [comment]
@ahrtr Have just rebased the PR and raised #16178. I was waiting for this to be merged first so didn't raise it previously. Sorry for misunderstanding what you meant.

BTW, are all errors supposed to be in [api/v3rpc/rpctypes/error.go](https://github.com/etcd-io/etcd/blob/89bbba8fb34bfff19fce330715e2ea8f5523cc8b/api/v3rpc/rpctypes/error.go)? I've just realized that Raft and corrupt uses HTTP rather than gRPC. I think using errors with the 'grpc' prefix is a bit confusing.

**Comment by ahrtr** [comment]
> I think using errors with the 'grpc' prefix is a bit confusing.

Right. We should avoid defining errors in `api/v3rpc/rpctypes/error.go` in this case. Instead, we should follow exactly the same pattern as `mvcc.ErrCompacted`

File: server/etcdserver/corrupt.go
Comment (ahrtr) [inline server/etcdserver/corrupt.go:525]:
```suggestion
		http.Error(w, mvcc.ErrClusterIdMismatch.Error(), http.StatusPreconditionFailed)
```

File: api/v3rpc/rpctypes/error.go
Comment (ahrtr) [reply api/v3rpc/rpctypes/error.go]: Or expose errClusterIDMismatch, and use it directly https://github.com/etcd-io/etcd/blob/693f25d6b98f5070527bc2b461f0992227c01c14/server/etcdserver/api/rafthttp/http.go#L57

File: server/etcdserver/corrupt.go
Comment (ahrtr) [reply server/etcdserver/corrupt.go:525]: Or expose errClusterIDMismatch, and use it directly https://github.com/etcd-io/etcd/blob/693f25d6b98f5070527bc2b461f0992227c01c14/server/etcdserver/api/rafthttp/http.go#L57

**Review by ahrtr — APPROVED** [review]
LGTM

Thanks @CaojiamingAlan

**Comment by serathius** [comment]
Was this backported to v3.4 and v3.5? @CaojiamingAlan can you backport?

**Pull request merged**
